File: Bridge/System/Console.cs
using Bridge;

namespace System {
    [External]
    public static class Console {
        public extern static int Read();
        public extern static string ReadLine();
        public extern static void Write(object v);

        [Template("System.Console.write(string.char({v}))")]
        public extern static void Write(char v);

        [Template("System.Console.write(System.String({buffer}))")]
        public extern static void Write(char[] buffer);

        [Template("System.Console.write(System.String({buffer}, {index}, {count}))")]
        public extern static void Write(char[] buffer, int index, int count);

        public extern static void Write(string format, params object[] arg);

        public extern static void WriteLine(object v);

        [Template("System.Console.writeLine(string.char({v}))")]
        public extern static void WriteLine(char v);

        [Template("System.Console.writeLine(System.String({buffer}))")]
        public extern static void WriteLine(char[] buffer);

        [Template("System.Console.writeLine(System.String({buffer}, {index}, {count}))")]
        public extern static void WriteLine(char[] buffer, int index, int count);

        public extern static void WriteLine(string format, params object[] arg);
    }
}

File: Example/__bin/coresystem.lua/CoreSystem/Console.lua
local System = System
local String = System.String
local format = String.format

local io = io
local stdin = io.stdin
local stdout = io.stdout
local read = stdin.read
local write = stdout.write
local tostring = tostring
local select = select
local byte = string.byte

local Console = {}

function Console.read()
    local ch = read(stdin, 1)
    return byte(ch)
end

function Console.readLine()
   return read(stdin)
end

function Console.write(v, ...)
    if select("#", ...) ~= 0 then
        v = format(v, ...)
    else
        v = tostring(v)      
    end
    write(stdout, v)     
end

function Console.writeLine(v, ...)
    if select("#", ...) ~= 0 then
        v = format(v, ...)
    else
        v = tostring(v)      
    end
    write(stdout, v, '\n')     
end

System.define("System.Console", Console)